# Fake news detector:

### Importing Packages:

In [1]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

### Importing data

In [20]:
file = pd.read_csv(r"C:\Users\USP\AppData\Local\Temp\Temp1_news.zip\news.csv")


print("This is our data\n")
print(file.shape)
file.head()

This is our data

(6335, 4)


Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

### Labelling 

In [21]:
labelling = file.label

labelling.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

### Splitting data

In [5]:
x_train, x_test, y_train, y_test = train_test_split(file['text'], labels, test_size=0.2, random_state=7)

In [6]:
x_train

6237    The head of a leading survivalist group has ma...
3722    ‹ › Arnaldo Rodgers is a trained and educated ...
5774    Patty Sanchez, 51, used to eat 13,000 calories...
336     But Benjamin Netanyahu’s reelection was regard...
3622    John Kasich was killing it with these Iowa vot...
4024    Email \nIt has become increasingly apparent th...
1537    Tesla, ‘World’s Safest Car,’ Explodes Like a B...
5200    For those of you salivating — or trembling — a...
279     American Express disowns Pink Floyd singer Rog...
4973    Osteoporosis (yes, magnesium is more important...
1283    Duterte Calls US Admin ‘Monkeys’ for Halting A...
4104    (CNN) Donald Trump dominated Super Tuesday, no...
876     Donald Trump has just five days until the Repu...
4769    Home / #Solutions / Colin Kaepernick Just Star...
5080    Black Turnout Soft in Early Voting, Boding Ill...
3601    Seven of the eight leading Republican presiden...
4722    \nOn Sunday evening, side-by-side images of Hi...
3561    Next P

In [7]:
x_test

3534    A day after the candidates squared off in a fi...
6265    VIDEO : FBI SOURCES SAY INDICTMENT LIKELY FOR ...
3123    It's debate season, where social media has bro...
3940    Mitch McConnell has decided to wager the Repub...
2856    Donald Trump, the actual Republican candidate ...
3031    Not all the claims in the vice presidential de...
4854    WASHINGTON—Top administration officials forcef...
5861    Top Dems want White House to call off Part B d...
307     Corrections and clarifications: An earlier ver...
2956    posted by Eddie Censorship by Facebook has bec...
4594    Email \n\nWSJ The political organization of Vi...
5268    On Sunday during an appearance on "Face the Na...
3929    After Russian President Vladimir Putin complim...
4276    Senate supporters of the Keystone XL pipeline,...
5655    The Republican presidential race continues to ...
5355    Wed, 26 Oct 2016 17:26 UTC © Sandro Perozzi/AP...
2268    (CNN) Matt Bevin, the controversial Kentuckian...
4525    Washin

In [8]:
y_train

6237    FAKE
3722    FAKE
5774    FAKE
336     REAL
3622    REAL
4024    FAKE
1537    FAKE
5200    REAL
279     FAKE
4973    FAKE
1283    FAKE
4104    REAL
876     REAL
4769    FAKE
5080    FAKE
3601    REAL
4722    FAKE
3561    FAKE
138     REAL
1949    FAKE
4675    FAKE
504     FAKE
4526    REAL
4848    FAKE
4741    REAL
3712    FAKE
2554    FAKE
913     REAL
2759    REAL
3131    FAKE
        ... 
3896    FAKE
1349    REAL
3775    REAL
3116    FAKE
3516    REAL
2298    REAL
2871    REAL
5195    FAKE
391     FAKE
6015    REAL
1456    REAL
1092    REAL
6118    FAKE
3239    FAKE
5000    FAKE
2778    REAL
4462    FAKE
4441    FAKE
3656    FAKE
2583    REAL
3257    REAL
1372    REAL
5735    REAL
919     FAKE
4307    FAKE
5699    FAKE
2550    REAL
537     REAL
1220    REAL
4271    REAL
Name: label, Length: 5068, dtype: object

In [9]:
y_test

3534    REAL
6265    FAKE
3123    REAL
3940    REAL
2856    REAL
3031    REAL
4854    REAL
5861    REAL
307     REAL
2956    FAKE
4594    FAKE
5268    REAL
3929    REAL
4276    REAL
5655    REAL
5355    FAKE
2268    REAL
4525    REAL
5979    REAL
1854    FAKE
2463    FAKE
4460    REAL
305     REAL
4736    FAKE
5098    REAL
351     REAL
4932    REAL
5477    FAKE
5370    FAKE
1328    FAKE
        ... 
1535    FAKE
1390    FAKE
444     FAKE
640     REAL
4800    FAKE
6196    REAL
1208    FAKE
459     FAKE
1598    REAL
5622    FAKE
1955    FAKE
1153    REAL
3202    FAKE
6089    FAKE
587     FAKE
625     FAKE
1675    REAL
3961    REAL
2646    FAKE
5544    REAL
1181    REAL
4568    FAKE
1953    REAL
735     FAKE
1137    FAKE
4986    REAL
5789    REAL
4338    REAL
5924    FAKE
6030    REAL
Name: label, Length: 1267, dtype: object

### Working with TfidfVectorizer

In [22]:

tfidf_vector=TfidfVectorizer(stop_words='english', max_df=0.7)


tfidf_train=tfidf_vector.fit_transform(x_train) 
tfidf_test=tfidf_vector.transform(x_test)

In [23]:
print(tfidf_train)

  (0, 25318)	0.06356197880227954
  (0, 31796)	0.032835312190384304
  (0, 53256)	0.08468714610538439
  (0, 24260)	0.02591230211547521
  (0, 49636)	0.04842197282925203
  (0, 3378)	0.044440694689591796
  (0, 57236)	0.04731169310146364
  (0, 51968)	0.04086068970720691
  (0, 23825)	0.06665750504864855
  (0, 8860)	0.06997151470751813
  (0, 48721)	0.07298886000547623
  (0, 10056)	0.08875931355939351
  (0, 56878)	0.22907566353089
  (0, 8953)	0.07617756364829195
  (0, 17858)	0.08179789731649102
  (0, 57811)	0.05018134592598926
  (0, 23481)	0.13387265628974546
  (0, 18378)	0.16712933698409224
  (0, 9851)	0.026344757209787056
  (0, 27283)	0.114537831765445
  (0, 4259)	0.1252399574231224
  (0, 19585)	0.06609677424160539
  (0, 58158)	0.03931037043190289
  (0, 8952)	0.06669214616557032
  (0, 15390)	0.18274094576335073
  :	:
  (5067, 25194)	0.06815122596851944
  (5067, 40823)	0.06020842915278987
  (5067, 51237)	0.06880620615191042
  (5067, 49605)	0.06329472428436525
  (5067, 37290)	0.0710310464405351

In [24]:
print(tfidf_test)

  (0, 60731)	0.05899712902382916
  (0, 60684)	0.033385466151529625
  (0, 60271)	0.04581143542258741
  (0, 60261)	0.07937859313949312
  (0, 59116)	0.10997273171965094
  (0, 59036)	0.08042180974421559
  (0, 58654)	0.07128159375531905
  (0, 58335)	0.0678398429566027
  (0, 57086)	0.12429244186413906
  (0, 55170)	0.20939665348422057
  (0, 54706)	0.035492943055135416
  (0, 54394)	0.10596727423829927
  (0, 54238)	0.06234899619642803
  (0, 53749)	0.04942070163765446
  (0, 53518)	0.117529167732626
  (0, 53144)	0.059982387365669215
  (0, 52555)	0.14065295472127948
  (0, 52483)	0.07360851972393109
  (0, 51960)	0.14065295472127948
  (0, 51955)	0.030414611451489323
  (0, 51663)	0.12057034351821985
  (0, 51527)	0.08134873077710283
  (0, 51159)	0.053278054236854326
  (0, 51005)	0.07360851972393109
  (0, 50068)	0.06481045119580665
  :	:
  (1266, 16835)	0.08080116269909657
  (1266, 16385)	0.0655727278454052
  (1266, 15999)	0.03279926109441314
  (1266, 14890)	0.05544599110567492
  (1266, 13110)	0.066971

### Initializing the PassiveAgressiveClassifier

In [16]:
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)


y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.82%


The accuracy is 92.82% in this model. To gain insights about the number of False and True negatives and positives, let's construct out confusionmatrix.

In [17]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[588,  50],
       [ 41, 588]], dtype=int64)